File: architecture-skills/design-patterns/examples/prototype/prototype.cs
// Prototype pattern in C#.
//
// Records make the pattern a language feature: `with` expressions produce
// nondestructive copies. Reference-typed members are still shared by the
// default copy, so a Clone method deep-copies them explicitly.

using System;
using System.Collections.Generic;

public sealed record Circle(double X, double Y, double Radius, string Color, List<string> Tags)
{
    // `with` copies references, so Tags would be aliased; Clone deep-copies it.
    public Circle Clone() => this with { Tags = new List<string>(Tags) };

    public string Describe() =>
        $"Circle r={Radius} {Color} at ({X}, {Y}) tags=[{string.Join(", ", Tags)}]";
}

/// <summary>Catalog of pre-configured prototypes, cloned instead of constructed.</summary>
public sealed class ShapeRegistry
{
    private readonly Dictionary<string, Circle> _prototypes = new();

    public void Register(string name, Circle prototypeShape) => _prototypes[name] = prototypeShape;

    public Circle Create(string name) => _prototypes[name].Clone();
}

public static class Program
{
    public static void Main()
    {
        var registry = new ShapeRegistry();
        registry.Register("warning-marker", new Circle(0, 0, 12, "red", new List<string> { "alert" }));

        Circle first = registry.Create("warning-marker");
        Circle second = registry.Create("warning-marker") with { X = 40, Y = 25 };
        second.Tags.Add("muted");

        Console.WriteLine(first.Describe());
        Console.WriteLine(second.Describe());
        Console.WriteLine($"independent copies: {!ReferenceEquals(first.Tags, second.Tags)}");
    }
}

File: architecture-skills/design-patterns/examples/prototype/prototype.rs
// Prototype pattern in Rust.
//
// #[derive(Clone)] is the pattern as a compiler-checked language feature:
// every field (including the Vec) is deep-copied. The registry shows the
// classic use — clone pre-configured exemplars instead of constructing.

use std::collections::HashMap;

#[derive(Clone, Debug)]
struct Circle {
    x: f64,
    y: f64,
    radius: f64,
    color: String,
    tags: Vec<String>,
}

impl Circle {
    fn describe(&self) -> String {
        format!(
            "Circle r={} {} at ({}, {}) tags={:?}",
            self.radius, self.color, self.x, self.y, self.tags
        )
    }
}

/// Catalog of pre-configured prototypes, cloned instead of constructed.
struct ShapeRegistry {
    prototypes: HashMap<String, Circle>,
}

impl ShapeRegistry {
    fn new() -> Self {
        Self {
            prototypes: HashMap::new(),
        }
    }

    fn register(&mut self, name: &str, prototype: Circle) {
        self.prototypes.insert(name.to_string(), prototype);
    }

    fn create(&self, name: &str) -> Circle {
        self.prototypes[name].clone()
    }
}

fn main() {
    let mut registry = ShapeRegistry::new();
    registry.register(
        "warning-marker",
        Circle {
            x: 0.0,
            y: 0.0,
            radius: 12.0,
            color: "red".to_string(),
            tags: vec!["alert".to_string()],
        },
    );

    let first = registry.create("warning-marker");
    let mut second = registry.create("warning-marker");
    second.x = 40.0;
    second.y = 25.0;
    second.tags.push("muted".to_string());

    println!("{}", first.describe());
    println!("{}", second.describe());
    // Ownership already guarantees the copies are independent: mutating
    // `second.tags` above could not have touched `first`.
}
